source/beamline/AMExclusiveStatesEnumeratedControl.cpp

```diff
@@ -0,0 +1,225 @@
+	QList<AMControl*> controls = indexControlMap_.values();
+
+	// If this control is connected, check whether any index controls that are set
+	// can stop. If there are no controls, this control returns false.
+
+	if (isConnected() && !controls.isEmpty()) {
+		bool canStop = true;
+
+		for (int i = 0, count = controls.count(); i < count && canStop; i++) {
+			AMControl *control = controls.at(i);
+			canStop = canStop && control && control->canStop();
+		}
+
+		result = canStop;
+	}
+
+	return result;
+}
+
+bool AMExclusiveStatesEnumeratedControl::setStatusControl(AMControl *newControl)
+{
+	bool result = false;
+
+	if (status_ != newControl) {
+
+		if (status_)
+			removeChildControl(status_);
+
+		status_ = newControl;
+
+		if (status_)
+			addChildControl(status_);
+
+		updateStates();
+
+		result = true;
+	}
+
+	return result;
+}
+
+void AMExclusiveStatesEnumeratedControl::updateConnected()
+{
+	// First, check to see if the status control is connected.
+
+	bool isConnected = status_ && status_->isConnected();
+
+	// If so, iterate through the controls of the added states and
+	// check their connected states. If there are no state controls,
+	// this control still reports as connected.
+
+	if (isConnected) {
+		QList<AMControl*> controls = indexControlMap_.values();
+
+		for (int i = 0, count = controls.count(); i < count && isConnected; i++) {
+			AMControl *control = controls.at(i);
+			isConnected = isConnected && control && control->isConnected();
+		}
+	}
+
+	setConnected(isConnected);
+}
+
+bool AMExclusiveStatesEnumeratedControl::addState(int index, const QString &stateName, double statusValue, AMControl *control, double controlTriggerValue)
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::addOption(index, stateName)) {
+		indexStatusMap_.insert(index, statusValue);
+		setControl(index, control);
+		indexTriggerMap_.insert(index, controlTriggerValue);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool AMExclusiveStatesEnumeratedControl::removeState(int index)
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::removeOption(index)) {
+		indexStatusMap_.remove(index);
+		removeControl(index);
+		indexTriggerMap_.remove(index);
+
+		result = true;
+	}
+
+	return result;
+}
+
+bool AMExclusiveStatesEnumeratedControl::clearStates()
+{
+	bool result = false;
+
+	if (AMEnumeratedControl::clearOptions()) {
+		indexStatusMap_.clear();
+		clearControls();
+		indexTriggerMap_.clear();
+
+		result = true;
+	}
+
+	return result;
+}
+
+AMAction3* AMExclusiveStatesEnumeratedControl::createMoveAction(double optionIndex)
+{
+	AMAction3 *result = 0;
+
+	// Match the given index setpoint with the corresponding control setpoint.
+	// Create and return action that moves the control to the control setpoint.
+	// We can assume that the optionIndex given is valid, according to the
+	// validSetpoint() provided.
+
+	AMControl *control = indexControlMap_.value(optionIndex, 0);
+	double trigger = indexTriggerMap_.value(optionIndex, -1);
```

Comment:
Since you have a default value of -1, would you want to check for that in your next if statement?